File: apps/desktop/src/ipc/commands.ts
import { invoke } from '@tauri-apps/api/core';

import type { CanvasNode, Connection, NodeKind, Size, Vec2, Viewport, Workspace } from './types';

/** Wrappers tipados dos comandos Tauri. Nenhum `invoke` solto no resto do app. */

export const workspaceLoad = (): Promise<Workspace> => invoke('workspace_load');

export const nodeCreate = (kind: NodeKind, position: Vec2, size: Size): Promise<CanvasNode> =>
  invoke('node_create', { kind, position, size });

/** Persiste um gesto inteiro e registra um passo unico de undo. */
export const nodesUpdate = (before: CanvasNode[], after: CanvasNode[]): Promise<void> =>
  invoke('nodes_update', { before, after });

export const nodeDelete = (id: string): Promise<void> => invoke('node_delete', { id });

export const nodeSetColor = (id: string, color: string): Promise<CanvasNode> =>
  invoke('node_set_color', { id, color });

// ---------------------------------------------------------------- workflows

export interface WorkflowSummary {
  id: string;
  name: string;
  rootPath: string;
}

export const workflowList = (): Promise<WorkflowSummary[]> => invoke('workflow_list');

export const workflowActive = (): Promise<string> => invoke('workflow_active');

export const workflowCreate = (name: string, rootPath: string): Promise<Workspace> =>
  invoke('workflow_create', { name, rootPath });

export const workflowActivate = (id: string): Promise<Workspace> =>
  invoke('workflow_activate', { id });

/** Roles customizadas, serializadas como JSON (o backend só guarda o blob). */
export const rolesLoad = (): Promise<string> => invoke('roles_load');
export const rolesSave = (json: string): Promise<void> => invoke('roles_save', { json });

/** Biblioteca de prompts, mesmo contrato de blob opaco das roles. */
export const promptsLoad = (): Promise<string> => invoke('prompts_load');
export const promptsSave = (json: string): Promise<void> => invoke('prompts_save', { json });

// ---------------------------------------------------------------- busca (M6)

export interface SearchHit {
  /** Caminho relativo à raiz do workflow. */
  path: string;
  /** Trecho recortado pelo FTS5, com o termo entre `[` e `]`. */
  snippet: string;
}

/** Reconstrói o índice do workflow ativo. Devolve quantos arquivos entraram. */
export const searchReindex = (): Promise<number> => invoke('search_reindex');

export const searchQuery = (query: string): Promise<SearchHit[]> =>
  invoke('search_query', { query });

/** Devolve `null` quando a aresta e recusada (laço, duplicata, nó inexistente). */
export const connectionCreate = (from: string, to: string): Promise<Connection | null> =>
  invoke('connection_create', { from, to });

export const connectionDelete = (id: string): Promise<void> => invoke('connection_delete', { id });

export const historyUndo = (): Promise<Workspace> => invoke('history_undo');

export const historyRedo = (): Promise<Workspace> => invoke('history_redo');

// ---------------------------------------------------------------- colaboração (M4)

export interface InboxCount {
  nodeId: string;
  pending: number;
}

export interface McpCall {
  nodeId: string;
  tool: string;
  ok: boolean;
  at: number;
}

/** Mensagens pendentes por agente, para o badge de inbox no canvas. */
export const agentInboxes = (): Promise<InboxCount[]> => invoke('agent_inboxes');

/** Chamadas MCP recentes — alimenta o debugger visual de agente. */
export const mcpActivity = (): Promise<McpCall[]> => invoke('mcp_activity');

/** Liga um orquestrador a vários workers de uma vez. Devolve as arestas criadas. */
export const connectMaestro = (orchestrator: string, workers: string[]): Promise<Connection[]> =>
  invoke('connect_maestro', { orchestrator, workers });

// ---------------------------------------------------------------- atenção

/** Espelho de `AgentStatus` (Rust, `attention.rs`). */
export type AttentionStatus = 'running' | 'waiting' | 'idle' | 'exited';

export interface AgentAttention {
  nodeId: string;
  status: AttentionStatus;
  /** Instante da última saída, em ms desde a epoch. */
  since: number;
}

/** Quem está produzindo e quem parou numa pergunta. Alimenta o anel do nó. */
export const agentAttention = (): Promise<AgentAttention[]> => invoke('agent_attention');

// ---------------------------------------------------------------- git

/** Espelho de `WorktreeStatus` (Rust, `git.rs`), achatado com o nó dono. */
export interface AgentWorktree {
  nodeId: string;
  branch: string;
  added: number;
  removed: number;
  /** Há mudança não commitada — `gitWorktreeMerge` só leva o que está commitado. */
  dirty: boolean;
}

/** A pasta do workflow ativo é um repositório git? Decide se o isolamento é oferecido. */
export const gitWorkflowIsRepo = (): Promise<boolean> => invoke('git_workflow_is_repo');

/** Cria o worktree do agente e devolve a pasta, para virar o `cwd` dele. */
export const gitWorktreeCreate = (name: string): Promise<string> =>
  invoke('git_worktree_create', { name });

export const gitAgentsStatus = (): Promise<AgentWorktree[]> => invoke('git_agents_status');

/** Integra a branch do agente no repositório principal. Devolve a saída do `git merge`. */
export const gitWorktreeMerge = (nodeId: string): Promise<string> =>
  invoke('git_worktree_merge', { nodeId });

/** Descarta o trabalho do agente: mata o processo e apaga worktree e branch. */
export const gitWorktreeDiscard = (nodeId: string): Promise<void> =>
  invoke('git_worktree_discard', { nodeId });

/** Um arquivo mexido, como o `GitNode` lista. */
export interface GitFile {
  path: string;
  /** Código do `git status --porcelain`: `M`, `A`, `D`, `R` ou `?` (não rastreado). */
  status: string;
  added: number;
  removed: number;
}

export interface GitNodeStatus {
  /** Agente dono do worktree, ou `null` quando o nó olha a raiz do workflow. */
  ownerId: string | null;
  branch: string;
  dirty: boolean;
  /** Só num worktree do Orkai integrar/descartar fazem sentido. */
  isWorktree: boolean;
  files: GitFile[];
}

/** Status do repositório que o `GitNode` observa — a conexão decide qual é. */
export const gitNodeStatus = (nodeId: string): Promise<GitNodeStatus> =>
  invoke('git_node_status', { nodeId });

export const gitNodeDiff = (nodeId: string, path: string): Promise<string> =>
  invoke('git_node_diff', { nodeId, path });

// ---------------------------------------------------------------- árvore de arquivos

export interface DirEntry {
  name: string;
  /** Caminho relativo à raiz do workflow, sempre com `/`. */
  path: string;
  isDir: boolean;
}

/** Conteúdo de uma pasta do workflow, um nível só (a árvore carrega o que expande). */
export const dirList = (path: string): Promise<DirEntry[]> => invoke('dir_list', { path });

export const viewportSave = (viewport: Viewport): Promise<void> =>
  invoke('viewport_save', { viewport });

export const fileRead = (path: string): Promise<string> => invoke('file_read', { path });

export const fileWrite = (path: string, content: string): Promise<void> =>
  invoke('file_write', { path, content });

export const ptySpawn = (nodeId: string, cols: number, rows: number): Promise<void> =>
  invoke('pty_spawn', { nodeId, cols, rows });

export const ptyWrite = (nodeId: string, data: string): Promise<void> =>
  invoke('pty_write', { nodeId, data });

export const ptyResize = (nodeId: string, cols: number, rows: number): Promise<void> =>
  invoke('pty_resize', { nodeId, cols, rows });

/** Historico da sessao em base64. */
export const ptyScrollback = (nodeId: string): Promise<string> =>
  invoke('pty_scrollback', { nodeId });

export const ptyKill = (nodeId: string): Promise<void> => invoke('pty_kill', { nodeId });

export const ptyDefaultShell = (): Promise<string> => invoke('pty_default_shell');

/** Se o binário do agente existe no PATH, pela mesma resolução que o spawn usa. */
export const commandAvailable = (command: string): Promise<boolean> =>
  invoke('command_available', { command });

File: apps/desktop/src-tauri/src/commands.rs
use std::path::PathBuf;

use base64::engine::general_purpose::STANDARD as BASE64;
use base64::Engine;
use orkai_core::{
    Command, Connection, ConnectionId, Node, NodeId, NodeKind, Size, Vec2, Viewport, Workspace,
};
use orkai_pty::{default_shell, PtyBackend, PtyEvent, PtySpec};
use orkai_storage::WorkflowSummary;
use serde::Serialize;
use tauri::{AppHandle, Emitter, State};
use uuid::Uuid;

use crate::error::{AppError, Result};
use crate::state::AppState;

#[derive(Clone, Serialize)]
#[serde(rename_all = "camelCase")]
struct PtyDataEvent {
    node_id: String,
    /// Bytes crus em base64: um bloco do PTY pode cortar um caractere UTF-8 ao meio,
    /// e converter para String aqui corromperia acentos e emoji na fronteira.
    data: String,
}

#[derive(Clone, Serialize)]
#[serde(rename_all = "camelCase")]
struct PtyExitEvent {
    node_id: String,
    code: i32,
}

// ---------------------------------------------------------------- workspace

#[tauri::command]
pub async fn workspace_load(state: State<'_, AppState>) -> Result<Workspace> {
    state.load_active().await
}

#[tauri::command]
pub async fn node_create(
    state: State<'_, AppState>,
    kind: NodeKind,
    position: Vec2,
    size: Size,
) -> Result<Node> {
    let workspace = state.load_active().await?;

    let node = Node {
        id: Uuid::new_v4(),
        kind,
        position: Vec2::new(position.x, position.y)?,
        size: Size::new(size.width, size.height)?,
        z_index: workspace.next_z_index(),
        created_at: agora_ms(),
    };

    // Um no markdown precisa do arquivo existindo para o editor abrir vazio em vez de falhar.
    if let NodeKind::Markdown { file_path, .. } = &node.kind {
        let caminho = state.resolve_path(file_path)?;
        if !caminho.exists() {
            if let Some(pai) = caminho.parent() {
                std::fs::create_dir_all(pai)?;
            }
            std::fs::write(&caminho, "")?;
        }
    }

    state.repo.save_node(state.active_id(), &node).await?;
    state.push_history(vec![Command::CreateNode { node: node.clone() }]);
    Ok(node)
}

/// Persiste mover/redimensionar em lote e registra um unico passo de undo.
///
/// O front chama no fim do gesto, nao a cada frame: um arrasto e um passo de undo,
/// e nao trezentos.
#[tauri::command]
pub async fn nodes_update(
    state: State<'_, AppState>,
    before: Vec<Node>,
    after: Vec<Node>,
) -> Result<()> {
    for node in &after {
        Vec2::new(node.position.x, node.position.y)?;
        Size::new(node.size.width, node.size.height)?;
    }

    let comando = Command::UpdateNodes { before, after };
    state.repo.apply(state.active_id(), &comando).await?;
    state.push_history(vec![comando]);
    Ok(())
}

#[tauri::command]
pub async fn node_delete(state: State<'_, AppState>, id: NodeId) -> Result<()> {
    let workspace = state.load_active().await?;
    let node = workspace.node(id)?.clone();

    // Ignora sessao inexistente: nem todo no e terminal.
    let _ = state.ptys.remove(id);
    // Descarta mensagens pendentes para o no que se foi.
    state.bus.forget(id);

    // As arestas viram comandos proprios no mesmo grupo: o undo as recria na ordem
    // certa, depois do no voltar a existir.
    let mut grupo: Vec<Command> = workspace
        .connections_of(id)
        .into_iter()
        .map(|connection| Command::DeleteConnection { connection })
        .collect();
    grupo.push(Command::DeleteNode { node });

    // O CASCADE do banco ja derruba as arestas; aplicar so o delete do no basta.
    state.repo.delete_node(id).await?;
    state.push_history(grupo);
    Ok(())
}

// ---------------------------------------------------------------- conexoes

#[tauri::command]
pub async fn connection_create(
    state: State<'_, AppState>,
    from: NodeId,
    to: NodeId,
) -> Result<Option<Connection>> {
    let mut workspace = state.load_active().await?;
    // Laco, duplicata e no inexistente sao recusados no dominio, nao aqui.
    let Some(connection) = workspace.add_connection(from, to) else {
        return Ok(None);
    };

    let comando = Command::CreateConnection {
        connection: connection.clone(),
    };
    state.repo.apply(state.active_id(), &comando).await?;
    state.push_history(vec![comando]);
    Ok(Some(connection))
}

#[tauri::command]
pub async fn connection_delete(state: State<'_, AppState>, id: ConnectionId) -> Result<()> {
    let mut workspace = state.load_active().await?;
    let connection = workspace.remove_connection(id)?;

    let comando = Command::DeleteConnection { connection };
    state.repo.apply(state.active_id(), &comando).await?;
    state.push_history(vec![comando]);
    Ok(())
}

// ---------------------------------------------------------------- historico

/// Desfaz o ultimo passo e devolve o workspace ja recarregado.
///
/// Devolver o estado inteiro em vez de um diff evita a classe de bug em que o front
/// aplica o inverso de um jeito e o banco de outro.
#[tauri::command]
pub async fn history_undo(state: State<'_, AppState>) -> Result<Workspace> {
    let comandos = state.history.lock().expect("historico envenenado").undo();
    state.aplicar_todos(comandos.unwrap_or_default()).await
}

#[tauri::command]
pub async fn history_redo(state: State<'_, AppState>) -> Result<Workspace> {
    let comandos = state.history.lock().expect("historico envenenado").redo();
    state.aplicar_todos(comandos.unwrap_or_default()).await
}

#[tauri::command]
pub async fn viewport_save(state: State<'_, AppState>, viewport: Viewport) -> Result<()> {
    state
        .repo
        .save_viewport(state.active_id(), &viewport)
        .await?;
    Ok(())
}

// ---------------------------------------------------------------- arquivos

#[tauri::command]
pub async fn file_read(state: State<'_, AppState>, path: PathBuf) -> Result<String> {
    let caminho = state.resolve_path(&path)?;
    Ok(std::fs::read_to_string(caminho).unwrap_or_default())
}

#[tauri::command]
pub async fn file_write(state: State<'_, AppState>, path: PathBuf, content: String) -> Result<()> {
    let caminho = state.resolve_path(&path)?;
    if let Some(pai) = caminho.parent() {
        std::fs::create_dir_all(pai)?;
    }
    std::fs::write(caminho, content)?;
    Ok(())
}

// ---------------------------------------------------------------- workflows

#[tauri::command]
pub async fn workflow_list(state: State<'_, AppState>) -> Result<Vec<WorkflowSummary>> {
    Ok(state.repo.list_workflows().await?)
}

/// Id do workflow ativo, para o front destacar na sidebar.
#[tauri::command]
pub fn workflow_active(state: State<'_, AppState>) -> String {
    state.active_id().to_string()
}

/// Cria um workflow ligado a uma pasta de projeto e ja o ativa.
#[tauri::command]
pub async fn workflow_create(
    state: State<'_, AppState>,
    name: String,
    root_path: String,
) -> Result<Workspace> {
    let nome = if name.trim().is_empty() {
        "Novo workflow"
    } else {
        name.trim()
    };
    let resumo = state.repo.create_workflow(nome, root_path.trim()).await?;
    state
        .set_active(resumo.id, PathBuf::from(&resumo.root_path))
        .await?;
    state.load_active().await
}

/// Troca o workflow ativo e devolve o novo, para o canvas recarregar.
#[tauri::command]
pub async fn workflow_activate(state: State<'_, AppState>, id: NodeId) -> Result<Workspace> {
    let resumo = state
        .repo
        .list_workflows()
        .await?
        .into_iter()
        .find(|w| w.id == id)
        .ok_or_else(|| AppError::Internal(format!("workflow inexistente: {id}")))?;
    state
        .set_active(resumo.id, PathBuf::from(&resumo.root_path))
        .await?;
    state.load_active().await
}

/// Muda a cor de uma nota. Devolve o no atualizado.
#[tauri::command]
pub async fn node_set_color(state: State<'_, AppState>, id: NodeId, color: String) -> Result<Node> {
    let workspace = state.load_active().await?;
    let mut node = workspace.node(id)?.clone();
    let NodeKind::Markdown { file_path, .. } = &node.kind else {
        return Err(AppError::Internal("cor so se aplica a notas".into()));
    };

    let antes = node.clone();
    node.kind = NodeKind::Markdown {
        file_path: file_path.clone(),
        color,
    };

    let comando = Command::UpdateNodes {
        before: vec![antes],
        after: vec![node.clone()],
    };
    state.repo.apply(state.active_id(), &comando).await?;
    state.push_history(vec![comando]);
    Ok(node)
}

// ---------------------------------------------------------------- roles

/// Chave em `app_setting` das roles customizadas (JSON opaco para o backend).
const CUSTOM_ROLES_KEY: &str = "custom_roles";

/// Roles criadas pelo usuario, como JSON. O backend so guarda o blob; o formato
/// (label + prompt) e concern do front.
#[tauri::command]
pub async fn roles_load(state: State<'_, AppState>) -> Result<String> {
    Ok(state
        .repo
        .get_setting(CUSTOM_ROLES_KEY)
        .await?
        .unwrap_or_else(|| "[]".into()))
}

#[tauri::command]
pub async fn roles_save(state: State<'_, AppState>, json: String) -> Result<()> {
    state.repo.set_setting(CUSTOM_ROLES_KEY, &json).await?;
    Ok(())
}

// ---------------------------------------------------------------- colaboracao (M4)

/// Contagem de mensagens pendentes por no, para o badge de inbox no canvas.
#[derive(Serialize)]
struct InboxCount {
    node_id: String,
    pending: usize,
}

#[tauri::command]
pub async fn agent_inboxes(state: State<'_, AppState>) -> Result<Vec<serde_json::Value>> {
    let workspace = state.load_active().await?;
    Ok(workspace
        .nodes
        .iter()
        .filter(|n| matches!(n.kind, NodeKind::Agent { .. }))
        .map(|n| {
            serde_json::to_value(InboxCount {
                node_id: n.id.to_string(),
                pending: state.bus.pending(n.id),
            })
            .expect("serializa")
        })
        .collect())
}

/// Chamadas MCP recentes — o debugger visual de agente ("veja cada tool call").
#[tauri::command]
pub fn mcp_activity(state: State<'_, AppState>) -> Vec<crate::mcp_server::McpCall> {
    state
        .mcp_log
        .lock()
        .expect("log envenenado")
        .iter()
        .cloned()
        .collect()
}

/// Modo Maestro: liga um orquestrador a varios workers de uma vez.
///
/// Cada aresta criada abre um canal MCP (a ACL do `orkai_send`). Reusa
/// `add_connection`, entao laco e duplicata sao recusados de graca. Devolve as arestas
/// realmente criadas, para o front atualizar sem recarregar tudo.
#[tauri::command]
pub async fn connect_maestro(
    state: State<'_, AppState>,
    orchestrator: NodeId,
    workers: Vec<NodeId>,
) -> Result<Vec<Connection>> {
    let mut workspace = state.load_active().await?;
    let mut criadas = Vec::new();

    for worker in workers {
        if let Some(connection) = workspace.add_connection(orchestrator, worker) {
            let comando = Command::CreateConnection {
                connection: connection.clone(),
            };
            state.repo.apply(state.active_id(), &comando).await?;
            criadas.push((connection, comando));
        }
    }

    // Um passo unico de undo desfaz o Maestro inteiro.
    if !criadas.is_empty() {
        state.push_history(criadas.iter().map(|(_, c)| c.clone()).collect());
    }
    Ok(criadas.into_iter().map(|(conn, _)| conn).collect())
}

// ---------------------------------------------------------------- pty

#[tauri::command]
pub async fn pty_spawn(
    app: AppHandle,
    state: State<'_, AppState>,
    node_id: NodeId,
    cols: u16,
    rows: u16,
) -> Result<()> {
    if state.ptys.contains(node_id) {
        return Ok(());
    }

    let workspace = state.load_active().await?;
    let node = workspace.node(node_id)?;
    let Some((comando, args, cwd)) = node.kind.spawn_spec() else {
        return Err(AppError::NotSpawnable(node_id.to_string()));
    };
    let eh_agente = matches!(node.kind, NodeKind::Agent { .. });

    // cwd vazio ou inexistente -> pasta do workflow ativo (a pasta do projeto).
    let cwd = if cwd.as_os_str().is_empty() || !cwd.exists() {
        state.active_root()
    } else {
        cwd.to_path_buf()
    };
    // Terminal manda comando vazio para pedir o shell padrao; agente sempre traz o seu.
    let comando = if comando.is_empty() {
        default_shell()
    } else {
        comando.to_string()
    };

    // Agente continua a conversa de onde parou: semeia o scrollback com o transcript
    // persistido. Terminal nao persiste, entao isto vem vazio.
    let seed = if eh_agente {
        state.repo.load_output(node_id).await?
    } else {
        Vec::new()
    };

    // Conecta o agente ao servidor MCP do Orkai: e assim que ele ganha as tools de
    // colaboracao (list_peers, send, inbox, read_peer_output). A URL carrega o id do no,
    // que e como o servidor sabe quem chamou e aplica a ACL.
    let mut args = args.to_vec();
    if eh_agente {
        args.extend(mcp_args(&comando, state.mcp_port, node_id));
    }

    let spec = PtySpec::new(comando, cwd)
        .with_args(args)
        .with_size(cols, rows)
        .with_seed(seed);
    let sessao = state.pty_backend.spawn(
        spec,
        std::sync::Arc::new(move |evento| match evento {
            PtyEvent::Data(bytes) => {
                let payload = PtyDataEvent {
                    node_id: node_id.to_string(),
                    data: BASE64.encode(&bytes),
                };
                if let Err(erro) = app.emit(&format!("pty://data/{node_id}"), payload) {
                    tracing::warn!(%erro, "falha ao emitir dados do terminal");
                }
            }
            PtyEvent::Exit { code } => {
                let payload = PtyExitEvent {
                    node_id: node_id.to_string(),
                    code,
                };
                if let Err(erro) = app.emit(&format!("pty://exit/{node_id}"), payload) {
                    tracing::warn!(%erro, "falha ao emitir saida do terminal");
                }
            }
        }),
    )?;

    // Persiste o transcript do agente periodicamente, para sobreviver ao restart.
    // Terminais sao efemeros: nao gastam escrita.
    if eh_agente {
        let scrollback = sessao.scrollback_arc();
        let repo = state.repo.clone();
        tauri::async_runtime::spawn(persistir_output(repo, node_id, scrollback));
    }

    state.ptys.insert(node_id, sessao);
    Ok(())
}

/// Salva o transcript do agente a cada intervalo enquanto a sessao viver.
///
/// A sessao segura uma referencia ao scrollback; quando ela e derrubada (no deletado),
/// `strong_count` cai e a tarefa faz o ultimo save e encerra. So grava quando mudou,
/// para nao reescrever o mesmo blob a cada tick depois que o processo termina.
async fn persistir_output(
    repo: orkai_storage::WorkspaceRepository,
    node_id: NodeId,
    scrollback: std::sync::Arc<std::sync::Mutex<orkai_pty::Scrollback>>,
) {
    use std::time::Duration;

    const INTERVALO: Duration = Duration::from_millis(500);
    let mut ultimo_len = usize::MAX;

    loop {
        tokio::time::sleep(INTERVALO).await;

        // 1 = so a propria tarefa segura o scrollback; a sessao ja morreu.
        let ultima_volta = std::sync::Arc::strong_count(&scrollback) == 1;

        let snapshot = scrollback.lock().expect("scrollback envenenado").snapshot();
        if snapshot.len() != ultimo_len {
            ultimo_len = snapshot.len();
            // Erro aqui e tolerado: se o no acabou de ser deletado, o CASCADE ja levou
            // a linha e a FK recusa o insert. Nao vale derrubar a tarefa por isso.
            if let Err(erro) = repo.save_output(node_id, &snapshot).await {
                tracing::debug!(%erro, "falha ao persistir transcript do agente");
            }
        }

        if ultima_volta {
            break;
        }
    }
}

#[tauri::command]
pub async fn pty_write(state: State<'_, AppState>, node_id: NodeId, data: String) -> Result<()> {
    let bytes = BASE64
        .decode(&data)
        .map_err(|e| AppError::Internal(e.to_string()))?;
    state.ptys.with(node_id, |sessao| sessao.write(&bytes))?;
    Ok(())
}

#[tauri::command]
pub async fn pty_resize(
    state: State<'_, AppState>,
    node_id: NodeId,
    cols: u16,
    rows: u16,
) -> Result<()> {
    state
        .ptys
        .with(node_id, |sessao| sessao.resize(cols, rows))?;
    Ok(())
}

/// Historico da sessao, para remontar o terminal apos a virtualizacao desmontar o no.
#[tauri::command]
pub async fn pty_scrollback(state: State<'_, AppState>, node_id: NodeId) -> Result<String> {
    let bytes = state.ptys.with(node_id, |sessao| Ok(sessao.scrollback()))?;
    Ok(BASE64.encode(bytes))
}

#[tauri::command]
pub async fn pty_kill(state: State<'_, AppState>, node_id: NodeId) -> Result<()> {
    let _ = state.ptys.remove(node_id);
    Ok(())
}

/// Shell padrao sugerido ao criar um terminal.
#[tauri::command]
pub fn pty_default_shell() -> String {
    default_shell()
}

/// Flags que conectam o CLI do agente ao servidor MCP do Orkai.
///
/// So o Claude Code esta cabeado aqui: aceita `--mcp-config` com JSON inline. Codex e
/// Aider registram MCP de outro jeito.
/// ponytail: um provider por vez; adiciona o proximo quando for testar com o binario real.
fn mcp_args(comando: &str, porta: u16, node_id: NodeId) -> Vec<String> {
    let url = crate::mcp_server::agent_url(porta, node_id);
    // Nome do binario sem caminho nem extensao: "C:/.../claude.exe" -> "claude".
    let base = std::path::Path::new(comando)
        .file_stem()
        .and_then(|s| s.to_str())
        .unwrap_or(comando)
        .to_ascii_lowercase();

    match base.as_str() {
        "claude" => {
            let config = serde_json::json!({
                "mcpServers": { "orkai": { "type": "http", "url": url } }
            });
            vec!["--mcp-config".to_string(), config.to_string()]
        }
        outro => {
            tracing::warn!(provider = outro, "MCP ainda nao cabeado para este agente");
            Vec::new()
        }
    }
}

fn agora_ms() -> i64 {
    std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .map(|d| d.as_millis() as i64)
        .unwrap_or(0)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn mcp_args_cabeia_claude_e_ignora_desconhecido() {
        let id = NodeId::new_v4();
        let args = mcp_args("claude", 7331, id);
        assert_eq!(args[0], "--mcp-config");
        assert!(args[1].contains(&id.to_string()), "url carrega o id do no");
        assert!(args[1].contains("\"type\":\"http\""));

        // Caminho completo e .exe tambem resolvem para claude.
        assert_eq!(mcp_args("C:/tools/claude.exe", 1, id).len(), 2);
        // Provider sem wiring: sem args, sem quebrar.
        assert!(mcp_args("codex", 7331, id).is_empty());
    }
}
